sentry_sdk/serializer.py

```diff
@@ -323,14 +343,108 @@ def _serialize_node_impl(
             if not isinstance(obj, string_types):
                 obj = safe_repr(obj)
 
+        # Allow span descriptions to be longer than other strings.
+        #
+        # For database auto-instrumented spans, the description contains
+        # potentially long SQL queries that are most useful when not truncated.
+        # Because arbitrarily large events may be discarded by the server as a
+        # protection mechanism, we dynamically limit the description length
+        # later in _truncate_span_descriptions.
+        if len(path) == 3 and path[0] == "spans" and path[-1] == "description":
+            span_description_bytes.append(len(obj))
+            return obj
         return _flatten_annotated(strip_string(obj))
 
+    def _truncate_span_descriptions(serialized_event, event, excess_bytes):
+        # type: (Event, Event, int) -> None
+        """
+        Modifies serialized_event in-place trying to remove excess_bytes from
+        span descriptions. The original event is used read-only to access the
+        span timestamps (represented as RFC3399-formatted strings in
+        serialized_event).
+
+        It uses heuristics to prioritize preserving the description of spans
+        that might be the most interesting ones in terms of understanding and
+        optimizing performance.
+        """
+        # When truncating a description, preserve a small prefix.
+        min_length = 10
+
+        def shortest_duration_longest_description_first(args):
+            # type: (Tuple[int, Span]) -> Tuple[timedelta, int]
+            i, serialized_span = args
+            span = event["spans"][i]
+            now = datetime.utcnow()
+            start = span.get("start_timestamp") or now
+            end = span.get("timestamp") or now
```

Comment:
Is `now` the most correct here? How is it even possible these can be nil by the time we hit the serializer?

I assume if we get a span without a valid start_timestamp or timestamp we should disregard it right?

Reply:
It is there for the sake of defensive coding. Is it not the role of the serializer to validate spans.

Note that `now` is not used to update the span, only to compute a `timedelta` for sorting.

Reply:
For sure, I see that. Maybe this is out of scope for the PR, but I wonder if we can make the guarantee that the spans are valid before hit hits the serializer. That way we don't need the extra func call.

Also, if `start` gets now, and `end` uses it's timestamp, wouldn't you get a negative timedelta?

Reply:
> Also, if `start` gets now, and `end` uses it's timestamp, wouldn't you get a negative timedelta?

The original start and end timestamps could also give a negative duration. That will only affect the sort order.